src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs

```diff
@@ -90,82 +91,100 @@ private void OnPrune(object? sender, PruningTriggerEventArgs e)
                 {
                     e.Status = PruningStatus.NotEnoughDiskSpace;
                 }
-                // we mark that we are waiting for block (for thread safety)
-                else if (Interlocked.CompareExchange(ref _waitingForBlockProcessed, 1, 0) == 0)
+                else
                 {
-                    // we don't want to start pruning in the middle of block processing, lets wait for new head.
-                    _blockTree.OnUpdateMainChain += OnUpdateMainChain;
                     e.Status = PruningStatus.Starting;
+
+                    _ = RunFullPruning(_processExitSource.Token);
                 }
             }
         }
 
-        private void OnUpdateMainChain(object? sender, OnUpdateMainChainArgs e)
+        private async Task WaitForMainChainChange(CancellationToken cancellationToken, Func<OnUpdateMainChainArgs, bool> handler)
```

Comment:
general convention is to put cancellation token at the last parameter (as it often has default)